snow/engine/snowman/bootstrap/context.go

```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package bootstrap
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+)
+
+var (
+	_ context.Context = (*haltableContext)(nil)
+
+	errHalted = errors.New("halted")
+)
+
+type haltableContext struct {
+	context.Context
+	common.Haltable
+}
+
+func (c *haltableContext) Err() error {
+	if c.Halted() {
+		return errHalted
+	}
+	return c.Context.Err()
+}
```

Comment:
This is a bit of an abstraction break... This function doesn't correctly implement the `context.Context` interface because it doesn't support `Done()` correctly... But that really isn't possible to implement correctly without modifying `common.Haltable` and introducing a new goroutine... I felt like this was a reasonable implementation (seeing as `common.Haltable` should really be removed anyway).